pkg/tempdir/tempdir.go

```diff
@@ -0,0 +1,306 @@
+type TempDir struct {
+	rootDir string
+
+	tempDirPath string
+	// tempDirLock is a lock file (e.g., rootDir/.lock-temp-dir-XYZ) specific to this
+	// TempDir instance, indicating it's in active use.
+	tempDirLock     *lockfile.LockFile
+	tempDirLockPath string
+}
+
+type staleDir struct {
+	lockPath    string
+	tempDirPath string
+}
+
+// listPotentialStaleDirs scans the rootDir for directories that might be stale temporary directories.
+// It identifies directories with the TempDirPrefix and their corresponding lock files with the TempdirLockPrefix.
+// If a directory does not have a corresponding lock file, it is considered stale.
+// The function returns a list of staleDir objects, each containing the paths to the directory and its lock file.
+func (td *TempDir) listPotentialStaleDirs() ([]staleDir, error) {
+	dirs := make(map[string]struct{})
+	locks := make(map[string]struct{})
+
+	dirContent, err := os.ReadDir(td.rootDir)
+	if err != nil {
+		return nil, fmt.Errorf("error reading temp dir %s: %w", td.rootDir, err)
+	}
+
+	for _, entry := range dirContent {
+		path := filepath.Join(td.rootDir, entry.Name())
+		if strings.HasPrefix(entry.Name(), TempDirPrefix) {
+			dirs[path] = struct{}{}
+			continue
+		}
+		if strings.HasPrefix(entry.Name(), TempdirLockPrefix) {
+			locks[path] = struct{}{}
+		}
+	}
+
+	out := []staleDir{}
+
+	for path := range dirs {
+		base := filepath.Base(path)
+		lockPath := filepath.Join(td.rootDir, TempdirLockPrefix+base)
+		if _, ok := locks[lockPath]; ok {
+			delete(locks, lockPath)
+		} else {
+			lockPath = ""
+		}
+		out = append(out, staleDir{
+			tempDirPath: path,
+			lockPath:    lockPath,
+		})
+	}
+
+	for path := range locks {
+		out = append(out, staleDir{
+			tempDirPath: "",
+			lockPath:    path,
+		})
+	}
+
+	return out, err
+}
+
+// RecoverStaleDirs identifies and removes stale temporary directories in the root directory.
+// A directory is considered stale if its lock file can be acquired (indicating no active use).
+// The function attempts to remove both the directory and its lock file.
+// Any errors encountered during the recovery process are collected and returned as a single error.
+func (td *TempDir) RecoverStaleDirs() error {
+	potentialStaleDirs, err := td.listPotentialStaleDirs()
+	if err != nil {
+		return fmt.Errorf("error listing potential stale temp dirs in %s: %w", td.rootDir, err)
+	}
+
+	if len(potentialStaleDirs) == 0 {
+		return nil
+	}
+
+	var recoveryErrors []error
+
+	for _, staleDir := range potentialStaleDirs {
+		if staleDir.lockPath != "" {
+			instanceLock, err := lockfile.GetLockFile(staleDir.lockPath)
+			recoveryErrors = append(recoveryErrors, fmt.Errorf("error getting lock file %s: %w", staleDir.lockPath, err))
+			if err := instanceLock.TryLock(); err != nil {
+				logrus.Debugf("Skipping cleanup for locked directory corresponding to %s", staleDir.lockPath)
+				continue
+			}
+		}
+		if staleDir.tempDirPath != "" {
```

Comment:
This is racy:

- Instance owner: create lock
- Recovery: `list…` finds a lock and no data dir
- Instance owner: create data dir
- Instance owner: crash
- Recovery: obtain lock
- Recovery: does not try to delete data

`list…` was called with no locks held, so it’s directory state is not authoritative. Only after `TryLock` succeeds we are the “owner” of the data, and we can check whether the data exists / try to delete it.